interp/constrintern.ml

```diff
@@ -2388,7 +2401,44 @@ let internalize globalenv env pattern_mode (_, ntnvars as lvar) c =
       (Id.Set.empty,Id.Map.empty,[]), None in
     (tm',(na.CAst.v, typ)), extra_id, match_td
 
-  and intern_impargs c env l subscopes args =
+  and intern_proj ?loc env expl (qid,us) args1 n args2 =
+    let args1, args1' = List.chop n args1 in
+    let c, args1 = List.sep_last args1 in
+    let args2 = args1'@args2 in
+    let f,args1 =
+      intern_applied_reference ~isproj:true intern env
+        (Environ.named_context_val globalenv) lvar us args1 qid
+    in
+    match find_projection_data f with
+    | Some (p, us, args0, nexpectedparams) ->
+      (* An official projection *)
+      check_not_notation_variable f ntnvars;
+      let impl, subscopes = find_appl_head_data env lvar f in
+      let imps1, imps2 =
+        if expl then
+          [], []
+        else
+          let ngivenparams = List.length (List.filter (fun (_,x) -> x == None) args1) in
+          let nextraargs = List.length args2 in
+          match select_impargs_size_for_proj ~nexpectedparams ~ngivenparams ~nextraargs impl with
+          | Inl (imps1,imps2) -> (imps1,imps2)
+          | Inr l ->
+            let l = Lazy.force l in
+            let n = match l with [n] -> n | _ -> 2 in (* singular only when l = [1] *)
+            user_err ?loc:qid.CAst.loc (str "Projection " ++ pr_qualid qid ++ str " expected " ++ pr_choice int l ++
+                           str (String.plural n " explicit parameter") ++ str ".")
+      in
+      let subscopes1, subscopes2 = List.chop (List.length args1 + 1) subscopes in
+      let c,args1 = List.sep_last (intern_impargs f env imps1 subscopes1 (args1@[c])) in
+      let p = DAst.make ?loc:qid.CAst.loc (GRef (GlobRef.ConstRef p,us)) in
+      let args2 = intern_impargs p env imps2 subscopes2 args2 in
+      smart_gapp p loc (args0@args1@c::args2)
+    | None ->
+      (* Tolerate a use of t.(f) notation for an ordinary application until a decision is taken about it *)
```

Comment:
This PR is making the `t.(p)` notation straying further than the application it used to stand for. Maybe we should take the decision *soon*. Note that the comment above about "officiality" is wrong, IIUC the code still takes the projection path when the head is an *arbitrary* reference.

Reply:
> This PR is making the t.(p) notation straying further than the application it used to stand for.

Yes and no. In this PR, it is still an application but with a treatment of implicit arguments which takes into account that the main argument splits the implicit arguments into two parts.

So, unless I'm missing something. nothing is changed wrt references which have no implicit arguments.

> Note that the comment above about "officiality" is wrong

Would replacing "An official projection" by "A reference registered as projection" be ok?

> IIUC the code still takes the projection path when the head is an arbitrary reference.

If by projection path you mean `intern_proj`, yes, since `intern_proj` is called whenever the `t.(f)` syntax is used. Is that ok or do you expect something else? For instance, would you prefer that another terminology than "proj" is used to talk about the `t.(f)` notation?

Reply:
@ppedrot: I used "A reference registered as projection", hoping it is compatible with your view. Should the names of functions (like `intern_proj`, etc.) be reconsidered further to avoid any ambiguity? Did I address your questions?

In any case, dependency is merged, will you assign?

[bedrock2's failure is waiting about mit-plv/bedrock2#190]